Analysis/EventFiltering/centralEventFilterProcessor.cxx

```diff
@@ -0,0 +1,185 @@
+      for (auto& filter : AvailableFilters) {
+        auto& filterConfig = config[filter];
+        if (config.IsObject()) {
+          std::unordered_map<std::string, float> tableMap;
+          for (auto& node : filterConfig.GetObject()) {
+            tableMap[node.name.GetString()] = node.value.GetDouble();
+            nCols++;
+          }
+          mDownscaling[filter] = tableMap;
+        }
+      }
+      break;
+    }
+  }
+  std::cout << "Middle init" << std::endl;
+  mScalers = new TH1D("mScalers", ";;Number of events", nCols + 1, -0.5, 0.5 + nCols);
+  mScalers->GetXaxis()->SetBinLabel(1, "Total number of events");
+
+  mFiltered = new TH1D("mFiltered", ";;Number of filtered events", nCols + 1, -0.5, 0.5 + nCols);
+  mFiltered->GetXaxis()->SetBinLabel(1, "Total number of events");
+
+  int bin{2};
+  for (auto& table : mDownscaling) {
+    for (auto& column : table.second) {
+      mScalers->GetXaxis()->SetBinLabel(bin, column.first.data());
+      mFiltered->GetXaxis()->SetBinLabel(bin++, column.first.data());
+    }
+  }
+
+  TFile test("test.root", "recreate");
+  mScalers->Clone()->Write();
+  test.Close();
+}
+
+void CentralEventFilterProcessor::run(ProcessingContext& pc)
+{
+  int64_t nEvents{-1};
+  for (auto& tableName : mDownscaling) {
+    auto tableConsumer = pc.inputs().get<TableConsumer>(tableName.first);
+
+    auto tablePtr{tableConsumer->asArrowTable()};
+    int64_t nRows{tablePtr->num_rows()};
+    nEvents = nEvents < 0 ? nRows : nEvents;
+    if (nEvents != nRows) {
+      std::cerr << "Inconsistent number of rows across trigger tables, fatal" << std::endl; ///TODO: move it to real fatal
+    }
+
+    auto schema{tablePtr->schema()};
+    for (auto& colName : tableName.second) {
+      int bin{mScalers->GetXaxis()->FindBin(colName.first.data())};
+      double binCenter{mScalers->GetXaxis()->GetBinCenter(bin)};
+      auto column{tablePtr->GetColumnByName(colName.first)};
+      double downscaling{colName.second};
+      if (column) {
+        for (int64_t iC{0}; iC < column->num_chunks(); ++iC) {
+          auto chunk{column->chunk(iC)};
+          auto boolArray = std::static_pointer_cast<arrow::BooleanArray>(chunk);
+          for (int64_t iS{0}; iS < chunk->length(); ++iS) {
+            if (boolArray->Value(iS)) {
+              mScalers->Fill(binCenter);
+              if (mUniformGenerator(mGeneratorEngine) < downscaling) {
+                mFiltered->Fill(binCenter);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  mScalers->SetBinContent(1, mScalers->GetBinContent(1) + nEvents);
+  mFiltered->SetBinContent(1, mFiltered->GetBinContent(1) + nEvents);
+}
+
+void CentralEventFilterProcessor::endOfStream(EndOfStreamContext& ec)
+{
+  TFile output("trigger.root", "recreate");
+  mScalers->Write("Scalers");
+  mFiltered->Write("FilteredScalers");
+  mScalers->Scale(1. / mScalers->GetBinContent(1));
+  mFiltered->Scale(1. / mFiltered->GetBinContent(1));
```

Comment:
protect for 0 entries?

Reply:
Do you mean the case of "mScalers->GetBinContent(1)" with 0 entries?

Reply:
Yes

Reply:
Done